lib/semantics/expression.cc

```diff
@@ -597,17 +597,25 @@ MaybeExpr ExpressionAnalyzer::Analyze(const parser::Name &n) {
   if (std::optional<int> kind{IsAcImpliedDo(n.source)}) {
     return AsMaybeExpr(ConvertToKind<TypeCategory::Integer>(
         *kind, AsExpr(ImpliedDoIndex{n.source})));
-  } else if (!context_.HasError(n)) {
+  } else if (context_.HasError(n)) {
+    return std::nullopt;
+  } else {
     const Symbol &ultimate{n.symbol->GetUltimate()};
     if (ultimate.detailsIf<semantics::TypeParamDetails>()) {
       // A bare reference to a derived type parameter (within a parameterized
       // derived type definition)
       return AsMaybeExpr(MakeBareTypeParamInquiry(&ultimate));
     } else {
-      return Designate(DataRef{ultimate});
+      const auto *details{ultimate.detailsIf<semantics::ObjectEntityDetails>()};
+      if (details && semantics::IsNamedConstant(ultimate) && details->type() &&
+          details->type()->AsIntrinsic() && !details->IsArray() &&
+          details->init()) {
+        return details->init();
```

Comment:
Wasn't evaluate `GetNamedConstantValue` already doing this replacement while folding expressions ? I guess not completely if you did it, but I am curious what.
The reason I am asking is that if `GetNamedConstantValue` was not yet called on ultimate, then conversion of the init expression to the parameter type may have not occurred yet. 
From a few tests, it looks OK, but it would probably be safer to add `details->initWasValidated()` in the check above.

Reply:
I did it this way because I found that if I just returned a DataRef to the original symbol (i.e. `k2` in `use m, k2 => k1`) the constant value was not propagated. But folding is a better place to take care of that, by changing `GetNamedConstantValue()` to follow the use-association.

@jeanPerier, @klausler, please take another look.

Reply:
Thanks for explaining and fixing `GetNamedConstantValue`, I had not even noticed it did not resolve use association...